File: binding/src/libqcommandlineoption.cpp
#include <QCommandLineOption>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qcommandlineoption.h>
#include "libqcommandlineoption.h"
#include "libqcommandlineoption.hxx"

QCommandLineOption* QCommandLineOption_new(const libqt_string name) {
	 return new QCommandLineOption(QString::fromUtf8(name.data, name.len));
}

QCommandLineOption* QCommandLineOption_new2(const libqt_list names) {
	 return new QCommandLineOption(QList<QString>());
}

QCommandLineOption* QCommandLineOption_new3(const libqt_string name, const libqt_string description) {
	 return new QCommandLineOption(QString::fromUtf8(name.data, name.len), QString::fromUtf8(description.data, description.len));
}

QCommandLineOption* QCommandLineOption_new4(const libqt_list names, const libqt_string description) {
	 return new QCommandLineOption(QList<QString>(), QString::fromUtf8(description.data, description.len));
}

QCommandLineOption* QCommandLineOption_new5(const QCommandLineOption* other) {
	 return new QCommandLineOption(*other);
}

QCommandLineOption* QCommandLineOption_new6(const libqt_string name, const libqt_string description, const libqt_string valueName) {
	 return new QCommandLineOption(QString::fromUtf8(name.data, name.len), QString::fromUtf8(description.data, description.len), QString::fromUtf8(valueName.data, valueName.len));
}

QCommandLineOption* QCommandLineOption_new7(const libqt_string name, const libqt_string description, const libqt_string valueName, const libqt_string defaultValue) {
	 return new QCommandLineOption(QString::fromUtf8(name.data, name.len), QString::fromUtf8(description.data, description.len), QString::fromUtf8(valueName.data, valueName.len), QString::fromUtf8(defaultValue.data, defaultValue.len));
}

QCommandLineOption* QCommandLineOption_new8(const libqt_list names, const libqt_string description, const libqt_string valueName) {
	 return new QCommandLineOption(QList<QString>(), QString::fromUtf8(description.data, description.len), QString::fromUtf8(valueName.data, valueName.len));
}

QCommandLineOption* QCommandLineOption_new9(const libqt_list names, const libqt_string description, const libqt_string valueName, const libqt_string defaultValue) {
	 return new QCommandLineOption(QList<QString>(), QString::fromUtf8(description.data, description.len), QString::fromUtf8(valueName.data, valueName.len), QString::fromUtf8(defaultValue.data, defaultValue.len));
}

void QCommandLineOption_OperatorAssign(QCommandLineOption* self, const QCommandLineOption* other) {
	self->operator=(*other);
}

void QCommandLineOption_Swap(QCommandLineOption* self, QCommandLineOption* other) {
	self->swap(*other);
}

libqt_list QCommandLineOption_Names(const QCommandLineOption* self) {
	auto _ret = self->names();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QCommandLineOption_SetValueName(QCommandLineOption* self, const libqt_string name) {
	self->setValueName(QString::fromUtf8(name.data, name.len));
}

libqt_string QCommandLineOption_ValueName(const QCommandLineOption* self) {
	QString _ret = self->valueName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCommandLineOption_SetDescription(QCommandLineOption* self, const libqt_string description) {
	self->setDescription(QString::fromUtf8(description.data, description.len));
}

libqt_string QCommandLineOption_Description(const QCommandLineOption* self) {
	QString _ret = self->description();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCommandLineOption_SetDefaultValue(QCommandLineOption* self, const libqt_string defaultValue) {
	self->setDefaultValue(QString::fromUtf8(defaultValue.data, defaultValue.len));
}

void QCommandLineOption_SetDefaultValues(QCommandLineOption* self, const libqt_list defaultValues) {
	self->setDefaultValues(QList<QString>());
}

libqt_list QCommandLineOption_DefaultValues(const QCommandLineOption* self) {
	auto _ret = self->defaultValues();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

int QCommandLineOption_Flags(const QCommandLineOption* self) {
	return self->flags();
}

void QCommandLineOption_SetFlags(QCommandLineOption* self, int aflags) {
	self->setFlags(static_cast<QFlags<QCommandLineOption::Flag>>(aflags));
}

void QCommandLineOption_Delete(QCommandLineOption* self) {
    delete self;
}

File: binding/include/libqcommandlineoption.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qcommandlineoption.html
pub const qcommandlineoption = struct {

    /// New constructs a new QCommandLineOption object.
    pub fn New(name: []const u8) QtC.QCommandLineOption {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QCommandLineOption_new(name_str);
    }


    /// New2 constructs a new QCommandLineOption object.
    pub fn New2(names: []const u8) QtC.QCommandLineOption {
        return qtc.QCommandLineOption_new2(@ptrCast(names));
    }


    /// New3 constructs a new QCommandLineOption object.
    pub fn New3(name: []const u8, description: []const u8) QtC.QCommandLineOption {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
return qtc.QCommandLineOption_new3(name_str, description_str);
    }


    /// New4 constructs a new QCommandLineOption object.
    pub fn New4(names: []const u8, description: []const u8) QtC.QCommandLineOption {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
return qtc.QCommandLineOption_new4(@ptrCast(names), description_str);
    }


    /// New5 constructs a new QCommandLineOption object.
    pub fn New5(other: ?*anyopaque) QtC.QCommandLineOption {
        return qtc.QCommandLineOption_new5(@ptrCast(other));
    }


    /// New6 constructs a new QCommandLineOption object.
    pub fn New6(name: []const u8, description: []const u8, valueName: []const u8) QtC.QCommandLineOption {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
const valueName_str = qtc.libqt_string{
    .len = valueName.len,
    .data = valueName.ptr,
};
return qtc.QCommandLineOption_new6(name_str, description_str, valueName_str);
    }


    /// New7 constructs a new QCommandLineOption object.
    pub fn New7(name: []const u8, description: []const u8, valueName: []const u8, defaultValue: []const u8) QtC.QCommandLineOption {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
const valueName_str = qtc.libqt_string{
    .len = valueName.len,
    .data = valueName.ptr,
};
const defaultValue_str = qtc.libqt_string{
    .len = defaultValue.len,
    .data = defaultValue.ptr,
};
return qtc.QCommandLineOption_new7(name_str, description_str, valueName_str, defaultValue_str);
    }


    /// New8 constructs a new QCommandLineOption object.
    pub fn New8(names: []const u8, description: []const u8, valueName: []const u8) QtC.QCommandLineOption {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
const valueName_str = qtc.libqt_string{
    .len = valueName.len,
    .data = valueName.ptr,
};
return qtc.QCommandLineOption_new8(@ptrCast(names), description_str, valueName_str);
    }


    /// New9 constructs a new QCommandLineOption object.
    pub fn New9(names: []const u8, description: []const u8, valueName: []const u8, defaultValue: []const u8) QtC.QCommandLineOption {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
const valueName_str = qtc.libqt_string{
    .len = valueName.len,
    .data = valueName.ptr,
};
const defaultValue_str = qtc.libqt_string{
    .len = defaultValue.len,
    .data = defaultValue.ptr,
};
return qtc.QCommandLineOption_new9(@ptrCast(names), description_str, valueName_str, defaultValue_str);
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCommandLineOption_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCommandLineOption_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn Names(self: ?*anyopaque, ) []const u8 {
        return qtc.QCommandLineOption_Names(@ptrCast(self));
    }

    pub fn SetValueName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QCommandLineOption_SetValueName(@ptrCast(self), name_str);
    }

    pub fn ValueName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineOption_ValueName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineoption.ValueName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetDescription(self: ?*anyopaque, description: []const u8) void {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
qtc.QCommandLineOption_SetDescription(@ptrCast(self), description_str);
    }

    pub fn Description(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineOption_Description(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineoption.Description: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetDefaultValue(self: ?*anyopaque, defaultValue: []const u8) void {
        const defaultValue_str = qtc.libqt_string{
    .len = defaultValue.len,
    .data = defaultValue.ptr,
};
qtc.QCommandLineOption_SetDefaultValue(@ptrCast(self), defaultValue_str);
    }

    pub fn SetDefaultValues(self: ?*anyopaque, defaultValues: []const u8) void {
        qtc.QCommandLineOption_SetDefaultValues(@ptrCast(self), @ptrCast(defaultValues));
    }

    pub fn DefaultValues(self: ?*anyopaque, ) []const u8 {
        return qtc.QCommandLineOption_DefaultValues(@ptrCast(self));
    }

    pub fn Flags(self: ?*anyopaque, ) i32 {
        return qtc.QCommandLineOption_Flags(@ptrCast(self));
    }

    pub fn SetFlags(self: ?*anyopaque, aflags: i32) void {
        qtc.QCommandLineOption_SetFlags(@ptrCast(self), @intCast(aflags));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCommandLineOption_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/commandlineoption.html#types
pub const enums = struct {
    pub const Flag = enum {
        pub const HiddenFromHelp: i32 = 1;
        pub const ShortOptionStyle: i32 = 2;
    };

};
